pin-project/src/lib.rs

```diff
@@ -0,0 +1,628 @@
+///
+/// ```rust
+/// # use pin_project::pin_projectable;
+/// # use std::pin::Pin;
+/// #[pin_projectable]
+/// struct Foo<T, U> {
+///     #[pin]
+///     future: T,
+///     field: U,
+/// }
+///
+/// impl<T, U> Foo<T, U> {
+///     fn baz(self: Pin<&mut Self>) {
+///         let this = self.project();
+///         let _: Pin<&mut T> = this.future;
+///         let _: &mut U = this.field;
+///     }
+/// }
+/// ```
+///
+/// ### Tuple structs (structs with unnamed fields):
+///
+/// ```rust
+/// # use pin_project::pin_projectable;
+/// # use std::pin::Pin;
+/// #[pin_projectable]
+/// struct Foo<T, U>(#[pin] T, U);
+///
+/// impl<T, U> Foo<T, U> {
+///     fn baz(self: Pin<&mut Self>) {
+///         let this = self.project();
+///         let _: Pin<&mut T> = this.0;
+///         let _: &mut U = this.1;
+///     }
+/// }
+/// ```
+///
+/// Structs without fields (unit-like struct and zero fields struct) are not
+/// supported.
+///
+/// ### Enums
+///
+/// `pin_projectable` also supports enums, but to use it ergonomically, you need
+/// to use the [`project`] attribute.
+///
+/// ```rust
+/// # #[cfg(feature = "project_attr")]
+/// use pin_project::{project, pin_projectable};
+/// # #[cfg(feature = "project_attr")]
+/// # use std::pin::Pin;
+///
+/// # #[cfg(feature = "project_attr")]
+/// #[pin_projectable]
+/// enum Foo<A, B, C> {
+///     Tuple(#[pin] A, B),
+///     Struct { field: C },
+///     Unit,
+/// }
+///
+/// # #[cfg(feature = "project_attr")]
+/// impl<A, B, C> Foo<A, B, C> {
+///     #[project] // Nightly does not need a dummy attribute to the function.
+///     fn baz(self: Pin<&mut Self>) {
+///         #[project]
+///         match self.project() {
+///             Foo::Tuple(x, y) => {
+///                 let _: Pin<&mut A> = x;
+///                 let _: &mut B = y;
+///             }
+///             Foo::Struct { field } => {
+///                 let _: &mut C = field;
+///             }
+///             Foo::Unit => {}
+///         }
+///     }
+/// }
+/// ```
+///
+/// Enums without variants (zero-variant enums) are not supported.
+///
+/// ### `#[pinned_drop]`
+///
+/// In order to correctly implement pin projections, a type's Drop impl must
+/// not move out of any stucturally pinned fields. Unfortunately, `Drop::drop` takes
+/// `&mut Self`, not `Pin<&mut Self>`.
+///
+/// To ensure that this requirement is upheld, the `pin_projectable` attribute will
+/// provide a `Drop` impl for you. This `Drop` impl will delegate to a function
+/// annotated with `#[pinned_drop]`, if present in the same `pin_project!` block.
+/// This function acts just like a normal `drop` impl, except for the fact that it
+/// takes `Pin<&mut Self>`. In particular, it will never be called more than once,
+/// just like `Drop::drop`.
+///
+/// For example:
+///
+/// ```rust
+/// use std::fmt::Debug;
+/// use pin_project::pin_project;
+/// use std::pin::Pin;
+///
+/// pin_project! {
+///     #[pin_projectable]
```

Comment:
Sounds good to me!

Reply:
How is this better than making `#[pin_project(Drop)]` generate a `Drop` impl in terms of a `PinnedDrop` impl, and then making `#[pinned_drop]` generate that `PinnedDrop` impl?

The private things should probably contain the word `Unsafe` or so, just to get across the message that misusing these can lead to unsoundness.

Reply:
@RalfJung
> How is this better than making `#[pin_project(Drop)]` generate a `Drop` impl in terms of a `PinnedDrop` impl, and then making `#[pinned_drop]` generate that `PinnedDrop` impl?

Uh... does that mean it makes `PinDrop` a public API with a `drop` method?

> The private things should probably contain the word `Unsafe` or so, just to get across the message that misusing these can lead to unsoundness.

 I agree. (And they should also be unsafe traits, as the name implies.)

Reply:
It makes it a "private" public API to be used only by the macro -- just like some other parts of the API.

Reply:
Thanks for your explanation. I think the way you mentioned is better for the following reasons.
* Macro is simple and easy to understand because it is probably only to rewrite the path.
* Error messages will be better, especially if the user wrote the method signature incorrectly.